flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java

```diff
@@ -395,12 +402,14 @@ public void onContainersAllocated(List<Container> containers) {
 						nodeManagerClient.startContainer(container, taskExecutorLaunchContext);
 					} catch (Throwable t) {
 						log.error("Could not start TaskManager in container {}.", container.getId(), t);
-
 						// release the failed container
 						workerNodeMap.remove(resourceId);
 						resourceManagerClient.releaseAssignedContainer(container.getId());
-						// and ask for a new one
-						requestYarnContainerIfRequired();
+						log.error("Could not start TaskManager in container {}.", container.getId(), t);
+						if (recordFailure()) {
+							// and ask for a new one
+							requestYarnContainerIfRequired();
+						}
```

Comment:
I think you are missing the `onContainersCompleted` case. If a container fails after the `startContainer` request has been sent, we won't record the failure and always allocate a new container. That's the reason why I would like to move this logic into the `ResourceManager` so that all container start call will go through the same code path.

Reply:
I think the changes are not properly tested. Otherwise `onContainersCompleted` should have been caught. Please add the relevant test cases.